### PR TITLE
hd derivation reward accounts

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#894  

# Overview

- [x] I have renamed ChangeChain to AccountingStyle, added MutableAccount constructor, and propagated chenges throught the codebase
- [x] Address derivation for MutableAccount works and checked

# Comments

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs [hunk 1]
```diff
@@ -121,25 +121,35 @@ spec = do
...
         it "An AddressPool always contains at least 'gap pool' addresses"
-            (property (prop_poolAtLeastGapAddresses @'InternalChain))
+            (property (prop_poolAtLeastGapAddresses @'UTxOInternal))
         it "Our addresses are eventually discovered"
-            (property (prop_poolEventuallyDiscoverOurs @'InternalChain))
+            (property (prop_poolEventuallyDiscoverOurs @'UTxOInternal))
+
+    describe "AddressPool MutableAccount" $ do
+        it "'lookupAddressPool' extends the pool by a maximum of 'gap'"
+            (checkCoverage (prop_poolGrowWithinGap @'MutableAccount))
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:147]: I guess this makes a lot of sense to test regardless, 👍 

but just wondering: are we going to use address pool with `MutableAccount`? Not now as we'll use a single reward account for each wallet (right?), but maybe sometime?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs [hunk 2]
```diff
@@ -158,32 +158,40 @@ import qualified Data.Text.Encoding as T
...
-    = ExternalChain
-    | InternalChain
+-- | Marker for addresses type engaged. We want to handle three cases here.
+-- The first two are pertinent to UTxO accounting
+-- and the last one handles rewards from participation in staking.
+-- (a) external chain is used for addresses that are part of the 'advertised'
+--     targets of a given transaction
+-- (b) internal change is for addresses used to handle the change of a
+--     the transaction within a given wallet
+-- (c) the addresses for a reward account
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs]: "Reward account" is just a specific case of an account though?

**Review by Anviking — APPROVED** [review]
Lgtm! (I might not be seeing the big picture here though) And nice commit history!

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs]: It is. @SebastienGllmt coined the term "Chimeric Account" recently to better describe the relationship with UTxO and the fact that such accounts actually co-exists with UTxOs.

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs [hunk 3]
```diff
@@ -189,9 +189,9 @@ prop_fingerprintInvalidAddress (InvalidAddress addr) = conjoin $ property <$>
                              Arbitrary Instances
 -------------------------------------------------------------------------------}
 
-instance Arbitrary ChangeChain where
+instance Arbitrary AccountingStyle where
     shrink _ = []
-    arbitrary = elements [InternalChain, ExternalChain]
+    arbitrary = elements [UTxOInternal, UTxOExternal]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs]: This should probably be `arbitraryBoundedEnum`. I don't see any particular reason to limit the properties to the first two levels only in this file.

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:147]: We'll have a single reward account indeed, and having those tests is debatable. It's a right step in the direction of supporting multiple chimeric accounts but it's not an immediate (and not even medium or long term) requirement. 

My recommendation would be not to have these, although, from a purely conceptual point of view, nothing's wrong with this.

**Review by KtorZ — APPROVED** [review]

File: lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs]: extended docs and added also `chimeric` word

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs]: indeed, used also `arbitraryBoundedEnum` in `Cardano.Wallet.Primitive.AddressDiscovery.SequentialSpec`

**Pull request merged**